Attachment: Use image element to show icon/thumbnail

squelart opened the pull request:
#### b0e7ef7b0420de70a7ec57e74412a2202718b9ca
<pre>
Attachment: Use image element to show icon/thumbnail
<a href="https://bugs.webkit.org/show_bug.cgi?id=258437">https://bugs.webkit.org/show_bug.cgi?id=258437</a>
rdar://105252742

Reviewed by Tim Nguyen.

Instead of using an inner attachment with legacy rendering, use a standard img element with a data URL
pointing at a blob that contains the image data.

Widespread changes:
- Removed m_innerLegacyAttachment and all related code.
- Removed image-only code.
- Consistently using &quot;wide-layout&quot; instead of &quot;modern&quot;, and &quot;narrow&quot; instead of &quot;legacy&quot;.

* LayoutTests/fast/attachment/mac/wide-attachment-image-controls-basic-expected.txt:
* LayoutTests/platform/ios-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* LayoutTests/platform/mac-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::didInsertAttachmentElement):
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::create):
(WebCore::HTMLAttachmentElement::didAddUserAgentShadowRoot):
(WebCore::attachmentPlaceholderIdentifier):
(WebCore::attachmentIconIdentifier):
(WebCore::HTMLAttachmentElement::ensureWideLayoutShadowTree):
(WebCore::HTMLAttachmentElement::updateProgress):
(WebCore::HTMLAttachmentElement::setFile):
(WebCore::HTMLAttachmentElement::attributeChanged):
(WebCore::HTMLAttachmentElement::attachmentActionForDisplay const):
(WebCore::HTMLAttachmentElement::attachmentTitleForDisplay const):
(WebCore::HTMLAttachmentElement::attachmentSubtitleForDisplay const):
(WebCore::HTMLAttachmentElement::updateAttributes):

(WebCore::HTMLAttachmentElement::updateImage):
This sets the img element with the correct &quot;src&quot; data contents (thumbnail or icon or nothing).

(WebCore::HTMLAttachmentElement::updateThumbnailForNarrowLayout):
(WebCore::HTMLAttachmentElement::updateThumbnailForWideLayout):
(WebCore::HTMLAttachmentElement::updateIconForNarrowLayout):
(WebCore::HTMLAttachmentElement::updateIconForWideLayout):
(WebCore::HTMLAttachmentElement::setNeedsWideLayoutIconRequest):
(WebCore::HTMLAttachmentElement::requestWideLayoutIconIfNeeded):
(WebCore::HTMLAttachmentElement::requestIconWithSize const):
(WebCore::attachmentPreviewIdentifier): Deleted.
(WebCore::HTMLAttachmentElement::ensureModernShadowTree): Deleted.
(WebCore::HTMLAttachmentElement::updateThumbnail): Deleted.
(WebCore::HTMLAttachmentElement::updateIcon): Deleted.
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/html/shadow/attachmentElementShadow.css:
(div#attachment-preview-area):
(img#attachment-icon):
(attachment#attachment-preview): Deleted.
* Source/WebCore/rendering/AttachmentLayout.mm:
(WebCore::AttachmentLayout::AttachmentLayout):
* Source/WebCore/rendering/RenderAttachment.cpp:

(WebCore::RenderAttachment::paintWideLayoutAttachmentOnly const):
Call requestWideLayoutIconIfNeeded when painting the shadow tree.

* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::attachmentIntrinsicSize const):
(WebCore::RenderThemeIOS::paintAttachment):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::paintAttachment):

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updateAttachmentThumbnail):
(WebKit::WebPage::updateAttachmentIcon):
Provide data blobs to wide-layout attachments.

Canonical link: <a href="https://commits.webkit.org/265615@main">https://commits.webkit.org/265615@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/36eb712032bd061018c53fc2b9b57e6d0d196b49

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11412 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11616 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11970 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13057 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10862 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11433 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14001 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11598 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13774 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11575 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12445 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9666 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13477 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9737 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10355 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17518 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10808 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10509 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13707 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10912 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8985 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10087 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2735 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14360 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10768 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -635,26 +604,78 @@ void HTMLAttachmentElement::updateEnclosingImageWithData(const String& contentTy
+        return;
+    }
+
+    m_imageElement->setSrc(nullAtom());
```
question: Is this going to display a broken image icon?

squelart replied:
It shows an empty square.

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -206,9 +206,9 @@ static Ref<ElementType> createContainedElement(HTMLElement& container, const Ato
 }
 
-void HTMLAttachmentElement::ensureModernShadowTree(ShadowRoot& root)
+void HTMLAttachmentElement::ensureWideLayoutShadowTree(ShadowRoot& root)
```
In terms of naming, I'd prefer we omit "WideLayout" from the new stuff, and add "Legacy" to the old stuff. We should be naming things assuming the legacy code is not there, so it's easy to remove later.

pxlcoder replied:
I think this will need to be done in a follow-up PR though, so we don't have a mix of names in the meantime.

nt1m replied:
We can also do renaming in a separate PR before this one and land this with more straightforward names, which I think is preferable over a follow up.

squelart replied:
The wide-layout UI is still not the default, it must be enabled through a preference, so I think it makes more sense for now to keep the naming as-is. The few name changes in this patch ("modern" -> "wide layout") just makes the code more consistent within itself and with the preference.
In the future, when we fully transition to the new UI everywhere and we get rid of the legacy UI, then it will make sense to drop the "wide layout" mentions.

nt1m replied:
It's confusing because the "Implementation" enum refers to "Legacy" vs. "WideLayout" which kind of implies the legacy one is going away.

squelart replied:
We actually don't know for sure that some apps won't want to keep it.
I will rename it to "NarrowLayout" (only used 4 times, small change). 🤔

nt1m commented on Source/WebCore/rendering/RenderAttachment.cpp:
```diff
@@ -79,6 +79,7 @@ void RenderAttachment::layout()
         RenderReplaced::layout();
         if (hasShadowContent())
             layoutShadowContent(intrinsicSize());
+        attachmentElement().requestWideLayoutIcon();
```
It would be better to call this from `HTMLAttachmentElement` (potentially `HTMLAttachmentElement::attributeChanged` ?).

Rendering code should be restricted to painting, not requesting content.

In fact, I think this is causing this function to be called too many times?

squelart replied:
`requestWideLayoutIcon()` is cheap to call multiple times, it only does a cross-process request when really necessary. Calling it during layout (as is done with legacy rendering) ensures that attributes have stabilized, otherwise there would be a risk of requesting an unnecessary icon.

But it may indeed be more appropriate to change this to an on-attribute-changed scheme, I'll think about it...

squelart replied:
Looking at the code again, I still think calling `requestWideLayoutIcon()` during painting may be the most efficient, it should only do one actual request in most cases, and none if the attachment is off-screen.
Otherwise we'd need to instrument at least `attributeChanged()`, but since we rely on two properties (`title` and `type`), that would mean sending two requests in almost all cases.

nt1m replied:
Hmm yeah this is tricky. Generally we do try to avoid changing the DOM from rendering code (which is sort what this is doing by mutating the image src attribute), since it creates cycles:
1. DOM is initialized
2. rendering/layout
3. DOM changes due to rendering
4. rendering/layout
etc.

I know there's a check to prevent the infinite cycle, but it would be good to avoid this pattern if you can find a way to call from `HTMLAttachmentElement`, not super important though.

squelart replied:
In this case, it wouldn't be an infinite cycle:
1. DOM init, attributes are changed and settled.
2. Rendering/layout (with empty img), triggering the icon request.
3. "Later", the icon arrives, img data is changed once.
4. Rendering/layout, hopefully limited to that fixed-sized img area?
5. The end.

Note that the icon request is asynchronous, so even if we sent it before the first rendering, the icon could still arrive later and trigger a second rendering anyway.

squelart replied:
Also we need the `renderRect` of the img when requesting the icon.

nt1m replied:
Yeah I know, `m_needsIconRequest` prevents this from being infinite, just saying the general "update DOM during rendering" pattern is something we try to avoid if it's low effort.

squelart replied:
One thing that may make me reconsider: Some ref tests seem a bit flaky, where I see either the actual or expected image is missing the icon! Moving the request a bit sooner _may_ help.

However, I'm not convinced this would be the right thing to do, since I think the real issue is that ref tests probably just take their snapshot at the end of the first full paint (right?), so as long as we have this asynchronous icon request, we'll always have the risk that the response could come too late and make the test fail; moving the request earlier would just make the failure less likely, not make it disappear completely.

So I would be more inclined in raising a separate radar about it, to solve this issue more broadly for all attachment tests (narrow/legacy and wide/modern). In case you know: Is there already a preferred way to delay reference tests taking their snapshot?

nt1m replied:
> In case you know: Is there already a preferred way to delay reference tests taking their snapshot?

Yes, `class="reftest-wait"` on the `<html>` tag lets you do so, then removing that class whenever the page is ready to be screenshot.

squelart replied:
Thanks. "Whenever the page is ready" is not trivial to know in this case: The icon is hidden inside a custom renderer or a UA shadow tree. 😅
I'm looking at adding an early request...

squelart replied:
Added an early request when attributes are set; there's a bit of complexity to avoid doing multiple requests if multiple attributes are set in one operation. There's still a potential icon request during rendering, in the unexpected case where no attributes are set.

squelart replied:
FYI, I did a quick test (without this patch) : in "onload", looping awaiting `requestAnimationFrame` until a test attachment element receives its icon, and it takes between 2 and 4 loops to get there! So I'm scared that all our attachment ref-tests are currently just comparing attachment bricks without icons 😱; intermittent failures probably happen when one side receives its icon unexpectedly quickly.
I'll prepare a proper bug separately... I don't think that should impact this patch here (as in, this icon-request race has been here for a long time, even before wide/modern attachments, so it shouldn't weigh on our decision to land this.)

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.h:
```diff
@@ -61,7 +60,9 @@ class HTMLAttachmentElement final : public HTMLElement {
     WEBCORE_EXPORT void updateThumbnail(const RefPtr<Image>& thumbnail);
+    WEBCORE_EXPORT void updateThumbnailImgSrcData(Vector<uint8_t>&&);
     WEBCORE_EXPORT void updateIcon(const RefPtr<Image>& icon, const WebCore::FloatSize&);
+    WEBCORE_EXPORT void updateIconImgSrcData(Vector<uint8_t>&&);
```
Ditto for naming here, I think we should just use `updateThumbnail` & `updateIcon` as function names, and change the old ones to be `updateLegacyThumbnail/Icon`.

squelart replied:
I think the naming is valuable to keep here: The argument is just a bag of bytes, so it's good to know that it's actually intended as a data blob for the `<img>`'s `src`.
Also, as above, the "old" UI is still the default.

nt1m replied:
I mean, `updateThumbnailImgSrcData` and `updateIconImgSrcData` end up calling `updateImage`, so from the user perspective, you are just updating the thumbnail and the icon.

squelart replied:
`updateImage` is private.

nt1m replied:
I guess my point is that these functions call `updateImage`, so they're not just updating the srcData fields, they're also updating the actual displayed image.

squelart replied:
These functions provide the thumbnail/icon data for a `src` field, then attachment *may* update the displayed image with that data.

Anyway, more importantly here I think: `WebPage` does a `if (wide) updateImgSrcData else update` and I've added assertions to verify that, so I now think that the names should reflect that precondition. Also I'm removing mentions of "legacy". So I'll change these functions to `Update{Narrow,Wide}{Thumbnail,Icon}`.

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.h:
```diff
@@ -128,7 +129,12 @@ class HTMLAttachmentElement final : public HTMLElement {
-    RefPtr<HTMLAttachmentElement> m_innerLegacyAttachment;
+    // The thumbnail is shown if non-empty, otherwise the icon is shown if non-empty.
+    Vector<uint8_t> m_thumbnailImgSrcData;
+    LayoutSize m_requestedIconSize; // {w<0||h<0} = no need to request, {0,0} = need request, {w>0,w>0} = requesting.
```
I don't think we need this three-state size (see my other comment for `requestWideLayoutIcon`). I'd just name this `m_iconSize`

squelart replied:
There's already an `m_iconSize`, it contains the size of the current icon (if any).
This `m_requestedIconSize` is what makes `requestWideLayoutIcon()` cheap to call multiple times, even while a request is in flight.

nt1m replied:
Can we replace this with a boolean so we don't duplicate the icon size? Something like `m_needsIconRequest`?

squelart replied:
Ok, since the img size is fixed, I guess we don't need the "currently requesting a specific size" state.

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.h:
```diff
@@ -128,7 +129,12 @@ class HTMLAttachmentElement final : public HTMLElement {
 
-    RefPtr<HTMLAttachmentElement> m_innerLegacyAttachment;
+    // The thumbnail is shown if non-empty, otherwise the icon is shown if non-empty.
+    Vector<uint8_t> m_thumbnailImgSrcData;
```
ditto for naming here. `m_thumbnail`

squelart replied:
Ditto

nt1m replied:
I guess for consistency we could have `m_{thumbnail/icon}For{Wide/Narrow}Layout`

The fact that it's a Vector or an Image is really a detail here.

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.h:
```diff
@@ -128,7 +129,12 @@ class HTMLAttachmentElement final : public HTMLElement {
+    // The thumbnail is shown if non-empty, otherwise the icon is shown if non-empty.
+    Vector<uint8_t> m_thumbnailImgSrcData;
+    LayoutSize m_requestedIconSize; // {w<0||h<0} = no need to request, {0,0} = need request, {w>0,w>0} = requesting.
+    Vector<uint8_t> m_iconImgSrcData;
```
same here.

squelart replied:
Same

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -635,26 +604,78 @@ void HTMLAttachmentElement::updateEnclosingImageWithData(const String& contentTy
+    if (!document().page() || !document().page()->attachmentElementClient())
+        return;
+    if (m_requestedIconSize.width() < 0.0)
+        return;
```
I think you shouldn't need this check if you only call `requestWideLayoutIcon` from `HTMLAttachmentElement` when needed.

nt1m commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7989,16 +7989,32 @@ void WebPage::updateAttachmentAttributes(const String& identifier, std::optional
     if (auto attachment = attachmentElementWithIdentifier(identifier)) {
-        if (RefPtr<ShareableBitmap> thumbnail = !qlThumbnailHandle.isNull() ? ShareableBitmap::create(WTFMove(qlThumbnailHandle)) : nullptr)
-            attachment->updateThumbnail(thumbnail->createImage());
+        if (RefPtr<ShareableBitmap> thumbnail = !qlThumbnailHandle.isNull() ? ShareableBitmap::create(WTFMove(qlThumbnailHandle)) : nullptr) {
```
Can we automatically deduce the type?

```suggestion
        if (RefPtr thumbnail = !qlThumbnailHandle.isNull() ? ShareableBitmap::create(WTFMove(qlThumbnailHandle)) : nullptr) {
```

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7989,16 +7989,32 @@ void WebPage::updateAttachmentAttributes(const String& identifier, std::optional
+        if (RefPtr<ShareableBitmap> thumbnail = !qlThumbnailHandle.isNull() ? ShareableBitmap::create(WTFMove(qlThumbnailHandle)) : nullptr) {
+            if (attachment->isWideLayout()) {
+                if (auto imageBuffer = ImageBuffer::create(thumbnail->size(), RenderingPurpose::Unspecified, 1.0, DestinationColorSpace::SRGB(), PixelFormat::BGRA8)) {
+                    thumbnail->paint(imageBuffer->context(), IntPoint(0, 0), IntRect(IntPoint(0, 0), thumbnail->size()));
```
`IntPoint::zero()`

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7989,16 +7989,32 @@ void WebPage::updateAttachmentAttributes(const String& identifier, std::optional
+        if (auto icon = !iconHandle.isNull() ? ShareableBitmap::create(WTFMove(iconHandle)) : nullptr) {
+            if (attachment->isWideLayout()) {
+                if (auto imageBuffer = ImageBuffer::create(icon->size(), RenderingPurpose::Unspecified, 1.0, DestinationColorSpace::SRGB(), PixelFormat::BGRA8)) {
+                    icon->paint(imageBuffer->context(), IntPoint(0, 0), IntRect(IntPoint(0, 0), icon->size()));
```
`IntPoint::zero()`

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.h:
```diff
@@ -60,8 +59,10 @@ class HTMLAttachmentElement final : public HTMLElement {
+    WEBCORE_EXPORT void updateNarrowThumbnail(const RefPtr<Image>& thumbnail);
+    WEBCORE_EXPORT void updateWideThumbnail(Vector<uint8_t>&&);
+    WEBCORE_EXPORT void updateNarrowIcon(const RefPtr<Image>& icon, const WebCore::FloatSize&);
+    WEBCORE_EXPORT void updateWideIcon(Vector<uint8_t>&&);
```
This is more of a nit, but I would prefer:

`update{Thumbnail/Image}For{Wide/Narrow}Layout`

The current names sort of imply a different meaning.

nt1m commented on Source/WebCore/html/shadow/attachmentElementShadow.css:
```diff
@@ -63,15 +63,15 @@ div#attachment-preview-area {
 #if (defined(WTF_PLATFORM_IOS_FAMILY) && WTF_PLATFORM_IOS_FAMILY)
-    max-width: 72px;
-    max-height: 72px;
+    width: 72px; /* Co-dependent with iconSize in HTMLAttachmentElement::needWideLayoutIcon */
+    height: 72px;
```
```suggestion
    width: 72px;
    height: 72px;
```

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.h:
```diff
@@ -140,6 +147,8 @@ class HTMLAttachmentElement final : public HTMLElement {
     RefPtr<HTMLElement> m_saveButton;
     mutable RefPtr<DOMRectReadOnly> m_saveButtonClientRect;
 
+    bool m_needWideLayoutIcon { false };
```
`m_needsWideLayoutIconRequest` ?

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -635,26 +608,80 @@ void HTMLAttachmentElement::updateEnclosingImageWithData(const String& contentTy
+    updateImage();
+}
+
+void HTMLAttachmentElement::needWideLayoutIcon()
```
```suggestion
void HTMLAttachmentElement::setNeedsWideLayoutIconRequest()
```

nt1m reviewed (approved):
Looks fine, I have naming nits for most part

squelart committed:
Attachment: Use image element to show icon/thumbnail
https://bugs.webkit.org/show_bug.cgi?id=258437
rdar://105252742

Reviewed by Tim Nguyen.

Instead of using an inner attachment with legacy rendering, use a standard img element with a data URL
pointing at a blob that contains the image data.

Widespread changes:
- Removed m_innerLegacyAttachment and all related code.
- Removed image-only code.
- Consistently using "wide-layout" instead of "modern", and "narrow" instead of "legacy".

* LayoutTests/fast/attachment/mac/wide-attachment-image-controls-basic-expected.txt:
* LayoutTests/platform/ios-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* LayoutTests/platform/mac-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::didInsertAttachmentElement):
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::create):
(WebCore::HTMLAttachmentElement::didAddUserAgentShadowRoot):
(WebCore::attachmentPlaceholderIdentifier):
(WebCore::attachmentIconIdentifier):
(WebCore::HTMLAttachmentElement::ensureWideLayoutShadowTree):
(WebCore::HTMLAttachmentElement::updateProgress):
(WebCore::HTMLAttachmentElement::setFile):
(WebCore::HTMLAttachmentElement::attributeChanged):
(WebCore::HTMLAttachmentElement::attachmentActionForDisplay const):
(WebCore::HTMLAttachmentElement::attachmentTitleForDisplay const):
(WebCore::HTMLAttachmentElement::attachmentSubtitleForDisplay const):
(WebCore::HTMLAttachmentElement::updateAttributes):

(WebCore::HTMLAttachmentElement::updateImage):
This sets the img element with the correct "src" data contents (thumbnail or icon or nothing).

(WebCore::HTMLAttachmentElement::updateThumbnailForNarrowLayout):
(WebCore::HTMLAttachmentElement::updateThumbnailForWideLayout):
(WebCore::HTMLAttachmentElement::updateIconForNarrowLayout):
(WebCore::HTMLAttachmentElement::updateIconForWideLayout):
(WebCore::HTMLAttachmentElement::setNeedsWideLayoutIconRequest):
(WebCore::HTMLAttachmentElement::requestWideLayoutIconIfNeeded):
(WebCore::HTMLAttachmentElement::requestIconWithSize const):
(WebCore::attachmentPreviewIdentifier): Deleted.
(WebCore::HTMLAttachmentElement::ensureModernShadowTree): Deleted.
(WebCore::HTMLAttachmentElement::updateThumbnail): Deleted.
(WebCore::HTMLAttachmentElement::updateIcon): Deleted.
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/html/shadow/attachmentElementShadow.css:
(div#attachment-preview-area):
(img#attachment-icon):
(attachment#attachment-preview): Deleted.
* Source/WebCore/rendering/AttachmentLayout.mm:
(WebCore::AttachmentLayout::AttachmentLayout):
* Source/WebCore/rendering/RenderAttachment.cpp:

(WebCore::RenderAttachment::paintWideLayoutAttachmentOnly const):
Call requestWideLayoutIconIfNeeded when painting the shadow tree.

* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::attachmentIntrinsicSize const):
(WebCore::RenderThemeIOS::paintAttachment):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::paintAttachment):

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updateAttachmentThumbnail):
(WebKit::WebPage::updateAttachmentIcon):
Provide data blobs to wide-layout attachments.

Canonical link: https://commits.webkit.org/265615@main
